torch/jit.py

```diff
@@ -1,26 +1,375 @@
+    a Module, parameters are automatically computed via `state_dict`).
+
+        >>> lstm = nn.LSTMCell(10, 20)
+        >>> @jit.trace(params=lstm.parameters())
+        >>> def f(a, b):
+        >>>     return lstm(a, b)
+    """
+    # TODO: handle decorating a class (not a callable)
+    def _trace(inner):
+        return TracedModule(inner, nderivs=nderivs, params=params)
+    if callable(arg):
+        return _trace(arg)
+    else:
+        return _trace
+
+
+# TODO: Formulating it this way means that state_dict of the traced thing
+# looks different, etc (because there's extra nesting).  Figure out a
+# way to avoid this
+class TracedModule(Module):
+    def __init__(self, inner, params=tuple(), nderivs=0):
+        super(TracedModule, self).__init__()
+        # inner may be a Module, or it may be an arbitrary callable
+        self.inner = inner
+        self.params = ParameterList(list(params))
+        self.nderivs = nderivs
+
+    def forward(self, *args):
+        # TODO: Possible optimization: use the unflattened
+        # output so we don't unflatten it when we get out
+        # NB: Not a method because trace_func_raw can't deal
+        # with methods
+        @raw_trace(nderivs=self.nderivs)
+        def traced_inner(in_vars, in_struct):
+            # NB: Uncommenting this line should be equivalent
+            # args, _params = _unflatten(in_vars, in_struct)
+            return _flatten(self.inner(*args))
+
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        trace, (out_vars, out_struct) = traced_inner(in_vars, in_struct)
+        out, extra = _unflatten(out_vars, out_struct)
+        assert len(extra) == 0
+        return trace, out
+
+
+# Functional version that assumes that all parameters are explicitly
+# specified
+def raw_trace(nderivs=0):
+    def _raw_trace(f):
+        @functools.wraps(f)
+        def wrapper(in_vars, in_struct=None):
+            trace = torch._C._tracer_enter(in_vars, nderivs)
+            out_vars, out_struct = f(in_vars, in_struct)
+            torch._C._tracer_exit(out_vars)
+            return trace, (out_vars, out_struct)
+        return wrapper
+    return _raw_trace
+
+
+# Lifecycle of a CompiledModule:
+# - It is given an underlying function, which knows how to actually
+#   execute the code that we want to compile.
+# - When we encounter an input configuration for which we don't
+#   have an optimized trace, we run the underlying function, tracing its
+#   result.  The trace is not done yet, so we save it into our set of pending
+#   traces for that configuration.
+# - When we encounter an input configuration whose trace is "ready"
+#   (that is, we've seen all of the passes, so the trace contains
+#   forwards/backwards/etc), we compile it, and then register this
+#   as the compiled trace.
+# - When we encounter an input configuration whose trace is compiled,
+#   we just directly run the compiled trace.
+class CompiledModule(Module):
```

Comment:
We should either monkey-patch the object, or implement all `__getattr__`, `__setattr__`, etc. such that you can still access the original object attributes without going through `inner`

Reply:
I agree that something should be done here. I am not exactly sure what the correct implementation strategy is.

* The trouble with monkeypatching is that it modifies the original model, which is a bad semantics for `torch.jit.compile()` (it would be less awful if Modules grew a `compile()` method that would inplace compile them.)
* The trouble with `__getattr__` is that it won't get triggered if there is a pre-existing method with a name already, and if you want `CompiledModule` to be a subclass of `Module` (this is a good idea, because many people, including us, `isinstance` check on `Module`).
* A straightforward way to implement this is to manually implement each special method in `Module` to forward to the inner class. It's mostly a question of making sure we get all the ones we need.
* If we really want to use `__getattr__`, we could manually delete the inherited methods, so they forward to `__getattr__`
* There's also `__getattribute__`, which truly overloads everything.

I kind of think the straightforward way might be best, even though it's very boilerplatey.